components/support/sql/src/conn_ext.rs

```diff
@@ -0,0 +1,122 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use rusqlite::{
+    self,
+    types::{ToSql, FromSql},
+    Connection,
+    Transaction,
+    Savepoint,
+    Row,
+    Result as SqlResult,
+};
+
+use maybe_cached::MaybeCached;
+
+/// This trait exists so that we can use these helpers on `rusqlite::{Transaction, Connection}`.
+/// Note that you must import ConnExt in order to call these methods on anything.
+pub trait ConnExt {
+
+    /// The method you need to implement to opt in to all of this.
+    fn conn(&self) -> &Connection;
+
+    /// Get a cached or uncached statement based on a flag.
+    fn prepare_maybe_cached<'conn>(&'conn self, sql: &str, cache: bool) -> SqlResult<MaybeCached<'conn>> {
+        MaybeCached::prepare(self.conn(), sql, cache)
+    }
+
+    /// Execute all the provided statements.
+    fn execute_all(&self, stmts: &[&str]) -> SqlResult<()> {
+        let conn = self.conn();
+        for sql in stmts {
+            conn.execute(sql, &[])?;
+        }
+        Ok(())
+    }
+
+    /// Equivalent to `Connection::execute_named` but caches the statement so that subsequent
+    /// calls to `execute_cached` will have imprroved performance.
+    fn execute_cached(&self, sql: &str, params: &[&dyn ToSql]) -> SqlResult<usize> {
+        let mut stmt = self.conn().prepare_cached(sql)?;
+        stmt.execute(params)
+    }
+
+    /// Equivalent to `Connection::execute_named` but caches the statement so that subsequent
+    /// calls to `execute_named_cached` will have imprroved performance.
+    fn execute_named_cached(&self, sql: &str, params: &[(&str, &dyn ToSql)]) -> SqlResult<usize> {
+        let mut stmt = self.conn().prepare_cached(sql)?;
+        stmt.execute_named(params)
+    }
+
+    /// Execute a query that returns a single result column, and return that result.
+    fn query_one<T: FromSql>(&self, sql: &str) -> SqlResult<T> {
+        let res: T = self.conn().query_row_and_then(sql, &[], |row| row.get_checked(0))?;
+        Ok(res)
+    }
+
+    /// Equivalent to `rusqlite::Connection::query_row_and_then` but allows use
+    /// of named parameters, and allows passing a flag to indicate that it's cached.
+    fn query_row_and_then_named<T, E, F>(
+        &self,
+        sql: &str,
+        params: &[(&str, &dyn ToSql)],
+        mapper: F,
+        cache: bool,
+    ) -> Result<T, E>
+    where
+        Self: Sized,
+        E: From<rusqlite::Error>,
+        F: FnOnce(&Row) -> Result<T, E>,
+    {
+        Ok(self.try_query_row(sql, params, mapper, cache)?
+            .ok_or(rusqlite::Error::QueryReturnedNoRows)?)
+    }
+
+    // This should probably have a longer name...
+    /// Like `query_row_and_then_named` but returns None instead of erroring if no such row exists.
+    fn try_query_row<T, E, F>(
```

Comment:
I think this is a fine name! Maybe `try_query_row_named`, if you want, but it's pretty clear what it does.